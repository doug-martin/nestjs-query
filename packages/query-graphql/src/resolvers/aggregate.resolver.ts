import { AggregateQuery, AggregateResponse, Class, Filter, mergeFilter, QueryService } from '@nestjs-query/core';
import { Args, ArgsType, Resolver } from '@nestjs/graphql';
import omit from 'lodash.omit';
import { AuthorizerInterceptor } from '../interceptors';
import { getDTONames } from '../common';
import { AggregateQueryParam, AuthorizerFilter, ResolverMethodOpts, ResolverQuery, SkipIf } from '../decorators';
import { AggregateArgsType, AggregateResponseType } from '../types';
import { transformAndValidate } from './helpers';
import { BaseServiceResolver, ResolverClass, ServiceResolver } from './resolver.interface';
import { OperationGroup } from '../auth';

export type AggregateResolverOpts = {
  enabled?: boolean;
} & ResolverMethodOpts;

export interface AggregateResolver<DTO, QS extends QueryService<DTO, unknown, unknown>>
  extends ServiceResolver<DTO, QS> {
  aggregate(
    filter: AggregateArgsType<DTO>,
    aggregateQuery: AggregateQuery<DTO>,
    authFilter?: Filter<DTO>,
  ): Promise<AggregateResponse<DTO>[]>;
}

/**
 * @internal
 * Mixin to add `read` graphql endpoints.
 */
export const Aggregateable =
  <DTO, QS extends QueryService<DTO, unknown, unknown>>(DTOClass: Class<DTO>, opts?: AggregateResolverOpts) =>
  <B extends Class<ServiceResolver<DTO, QS>>>(BaseClass: B): Class<AggregateResolver<DTO, QS>> & B => {
    const { baseNameLower } = getDTONames(DTOClass);
    const commonResolverOpts = omit(opts, 'dtoName', 'one', 'many', 'QueryArgs', 'Connection');
    const queryName = `${baseNameLower}Aggregate`;
    const AR = AggregateResponseType(DTOClass);
    @ArgsType()
    class AA extends AggregateArgsType(DTOClass) {}

    @Resolver(() => AR, { isAbstract: true })
    class AggregateResolverBase extends BaseClass {
      @SkipIf(
        () => !opts || !opts.enabled,
        ResolverQuery(
          () => [AR],
          { name: queryName },
          commonResolverOpts,
          { interceptors: [AuthorizerInterceptor(DTOClass)] },
          opts ?? {},
        ),
      )
      async aggregate(
        @Args() args: AA,
        @AggregateQueryParam() query: AggregateQuery<DTO>,
        @AuthorizerFilter({
          operationGroup: OperationGroup.AGGREGATE,
          many: true,
        })
        authFilter?: Filter<DTO>,
      ): Promise<AggregateResponse<DTO>[]> {
        const qa = await transformAndValidate(AA, args);
        return this.service.aggregate(mergeFilter(qa.filter || {}, authFilter ?? {}), query);
      }
    }
    return AggregateResolverBase;
  };
// eslint-disable-next-line @typescript-eslint/no-redeclare -- intentional
export const AggregateResolver = <
  DTO,
  QS extends QueryService<DTO, unknown, unknown> = QueryService<DTO, unknown, unknown>,
>(
  DTOClass: Class<DTO>,
  opts?: AggregateResolverOpts,
): ResolverClass<DTO, QS, AggregateResolver<DTO, QS>> => Aggregateable(DTOClass, opts)(BaseServiceResolver);
