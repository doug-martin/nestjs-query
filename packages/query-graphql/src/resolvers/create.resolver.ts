/**
 * This is the doc comment for file1.ts
 * @packageDocumentation
 */
// eslint-disable-next-line max-classes-per-file
import { Class, DeepPartial, Filter, QueryService } from '@nestjs-query/core';
import { Args, ArgsType, InputType, PartialType, Resolver } from '@nestjs/graphql';
import omit from 'lodash.omit';
import { HookTypes } from '../hooks';
import { DTONames, getDTONames } from '../common';
import { AuthorizerFilter, MutationHookArgs, ResolverMutation, ResolverSubscription } from '../decorators';
import { AuthorizerInterceptor, HookInterceptor } from '../interceptors';
import { EventType, getDTOEventName } from '../subscription';
import {
  CreateManyInputType,
  CreateOneInputType,
  MutationArgsType,
  SubscriptionArgsType,
  SubscriptionFilterInputType,
} from '../types';
import { createSubscriptionFilter, getSubscriptionEventName } from './helpers';
import { BaseServiceResolver, ResolverClass, ServiceResolver, SubscriptionResolverOpts } from './resolver.interface';
import { OperationGroup } from '../auth';

export type CreatedEvent<DTO> = { [eventName: string]: DTO };

export interface CreateResolverOpts<DTO, C = DeepPartial<DTO>> extends SubscriptionResolverOpts {
  /**
   * The Input DTO that should be used to create records.
   */
  CreateDTOClass?: Class<C>;
  /**
   * The class to be used for `createOne` input.
   */
  CreateOneInput?: Class<CreateOneInputType<C>>;
  /**
   * The class to be used for `createMany` input.
   */
  CreateManyInput?: Class<CreateManyInputType<C>>;

  createOneMutationName?: string;
  createManyMutationName?: string;
}

export interface CreateResolver<DTO, C, QS extends QueryService<DTO, C, unknown>> extends ServiceResolver<DTO, QS> {
  createOne(input: MutationArgsType<CreateOneInputType<C>>, authorizeFilter?: Filter<DTO>): Promise<DTO>;

  createMany(input: MutationArgsType<CreateManyInputType<C>>, authorizeFilter?: Filter<DTO>): Promise<DTO[]>;

  createdSubscription(
    input?: SubscriptionArgsType<DTO>,
    authorizeFilter?: Filter<DTO>,
  ): AsyncIterator<CreatedEvent<DTO>>;
}

/** @internal */
const defaultCreateDTO = <DTO, C>(dtoNames: DTONames, DTOClass: Class<DTO>): Class<C> => {
  @InputType(`Create${dtoNames.baseName}`)
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  class PartialInput extends PartialType(DTOClass, InputType) {}

  return PartialInput as Class<C>;
};

/** @internal */
const defaultCreateOneInput = <C>(dtoNames: DTONames, InputDTO: Class<C>): Class<CreateOneInputType<C>> => {
  const { baseName, baseNameLower } = dtoNames;
  @InputType(`CreateOne${baseName}Input`)
  class CO extends CreateOneInputType(baseNameLower, InputDTO) {}
  return CO;
};

/** @internal */
const defaultCreateManyInput = <C>(dtoNames: DTONames, InputDTO: Class<C>): Class<CreateManyInputType<C>> => {
  const { pluralBaseName, pluralBaseNameLower } = dtoNames;
  @InputType(`CreateMany${pluralBaseName}Input`)
  class CM extends CreateManyInputType(pluralBaseNameLower, InputDTO) {}
  return CM;
};

/**
 * @internal
 * Mixin to add `create` graphql endpoints.
 */
export const Creatable =
  <DTO, C, QS extends QueryService<DTO, C, unknown>>(DTOClass: Class<DTO>, opts: CreateResolverOpts<DTO, C>) =>
  <B extends Class<ServiceResolver<DTO, QS>>>(BaseClass: B): Class<CreateResolver<DTO, C, QS>> & B => {
    const dtoNames = getDTONames(DTOClass, opts);
    const { baseName, pluralBaseName } = dtoNames;
    const enableSubscriptions = opts.enableSubscriptions === true;
    const enableOneSubscriptions = opts.one?.enableSubscriptions ?? enableSubscriptions;
    const enableManySubscriptions = opts.many?.enableSubscriptions ?? enableSubscriptions;
    const createdEvent = getDTOEventName(EventType.CREATED, DTOClass);
    const {
      CreateDTOClass = defaultCreateDTO(dtoNames, DTOClass),
      CreateOneInput = defaultCreateOneInput(dtoNames, CreateDTOClass),
      CreateManyInput = defaultCreateManyInput(dtoNames, CreateDTOClass),
    } = opts;
    const createOneMutationName = opts.one?.name ?? `createOne${baseName}`;
    const createManyMutationName = opts.many?.name ?? `createMany${pluralBaseName}`;
    const commonResolverOpts = omit(
      opts,
      'dtoName',
      'one',
      'many',
      'CreateDTOClass',
      'CreateOneInput',
      'CreateManyInput',
    );

    @ArgsType()
    class CO extends MutationArgsType(CreateOneInput) {}

    @ArgsType()
    class CM extends MutationArgsType(CreateManyInput) {}

    @InputType(`Create${baseName}SubscriptionFilterInput`)
    class SI extends SubscriptionFilterInputType(DTOClass) {}

    @ArgsType()
    class SA extends SubscriptionArgsType(SI) {}

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const subscriptionFilter = createSubscriptionFilter(SI, createdEvent);

    @Resolver(() => DTOClass, { isAbstract: true })
    class CreateResolverBase extends BaseClass {
      @ResolverMutation(
        () => DTOClass,
        { name: createOneMutationName },
        commonResolverOpts,
        {
          interceptors: [
            HookInterceptor(HookTypes.BEFORE_CREATE_ONE, CreateDTOClass, DTOClass),
            AuthorizerInterceptor(DTOClass),
          ],
        },
        opts.one ?? {},
      )
      async createOne(
        @MutationHookArgs() input: CO,
        @AuthorizerFilter({
          operationGroup: OperationGroup.CREATE,
          many: false,
        }) // eslint-disable-next-line @typescript-eslint/no-unused-vars
        authorizeFilter?: Filter<DTO>,
      ): Promise<DTO> {
        // Ignore `authorizeFilter` for now but give users the ability to throw an UnauthorizedException
        const created = await this.service.createOne(input.input.input);
        if (enableOneSubscriptions) {
          await this.publishCreatedEvent(created, authorizeFilter);
        }
        return created;
      }

      @ResolverMutation(
        () => [DTOClass],
        { name: createManyMutationName },
        { ...commonResolverOpts },
        {
          interceptors: [
            HookInterceptor(HookTypes.BEFORE_CREATE_MANY, CreateDTOClass, DTOClass),
            AuthorizerInterceptor(DTOClass),
          ],
        },
        opts.many ?? {},
      )
      async createMany(
        @MutationHookArgs() input: CM,

        @AuthorizerFilter({
          operationGroup: OperationGroup.CREATE,
          many: true,
        }) // eslint-disable-next-line @typescript-eslint/no-unused-vars
        authorizeFilter?: Filter<DTO>,
      ): Promise<DTO[]> {
        // Ignore `authorizeFilter` for now but give users the ability to throw an UnauthorizedException
        const created = await this.service.createMany(input.input.input);
        if (enableManySubscriptions) {
          await Promise.all(created.map((c) => this.publishCreatedEvent(c, authorizeFilter)));
        }
        return created;
      }

      async publishCreatedEvent(dto: DTO, authorizeFilter?: Filter<DTO>): Promise<void> {
        if (this.pubSub) {
          const eventName = getSubscriptionEventName(createdEvent, authorizeFilter);
          await this.pubSub.publish(eventName, { [createdEvent]: dto });
        }
      }

      @ResolverSubscription(() => DTOClass, { name: createdEvent, filter: subscriptionFilter }, commonResolverOpts, {
        enableSubscriptions: enableOneSubscriptions || enableManySubscriptions,
        interceptors: [AuthorizerInterceptor(DTOClass)],
      })
      createdSubscription(
        @Args() input?: SA,
        @AuthorizerFilter({ operationGroup: OperationGroup.CREATE, many: false })
        authorizeFilter?: Filter<DTO>,
      ): AsyncIterator<CreatedEvent<DTO>> {
        if (!this.pubSub || !(enableManySubscriptions || enableOneSubscriptions)) {
          throw new Error(`Unable to subscribe to ${createdEvent}`);
        }

        const eventName = getSubscriptionEventName(createdEvent, authorizeFilter);
        return this.pubSub.asyncIterator<CreatedEvent<DTO>>(eventName);
      }
    }
    return CreateResolverBase;
  };

/**
 * Factory to create a new abstract class that can be extended to add `create` endpoints.
 *
 * Assume we have `TodoItemDTO`, you can create a resolver with `createOneTodoItem` and `createManyTodoItems` graphql
 * query endpoints using the following code.
 *
 * ```ts
 * @Resolver()
 * export class TodoItemResolver extends CreateResolver(TodoItemDTO) {
 *   constructor(readonly service: TodoItemService) {
 *    super(service);
 *   }
 * }
 * ```
 *
 * @param DTOClass - The DTO class that should be returned from the `createOne` and `createMany` endpoint.
 * @param opts - Options to customize endpoints.
 * @typeparam DTO - The type of DTO that should be created.
 * @typeparam C - The create DTO type.
 */
// eslint-disable-next-line @typescript-eslint/no-redeclare -- intentional
export const CreateResolver = <
  DTO,
  C = DeepPartial<DTO>,
  QS extends QueryService<DTO, C, unknown> = QueryService<DTO, C, unknown>,
>(
  DTOClass: Class<DTO>,
  opts: CreateResolverOpts<DTO, C> = {},
): ResolverClass<DTO, QS, CreateResolver<DTO, C, QS>> => Creatable(DTOClass, opts)(BaseServiceResolver);
