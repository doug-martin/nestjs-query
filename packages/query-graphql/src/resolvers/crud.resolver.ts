import { Class, DeepPartial, QueryService } from '@nestjs-query/core';
import { ConnectionOptions, PagingStrategies } from '../types';
import { Aggregateable, AggregateResolverOpts, AggregateResolver } from './aggregate.resolver';
import { Relatable } from './relations';
import { Readable, ReadResolverFromOpts, ReadResolverOpts } from './read.resolver';
import { Creatable, CreateResolver, CreateResolverOpts } from './create.resolver';
import { Referenceable, ReferenceResolverOpts } from './reference.resolver';
import { MergePagingStrategyOpts, ResolverClass } from './resolver.interface';
import { Updateable, UpdateResolver, UpdateResolverOpts } from './update.resolver';
import { DeleteResolver, DeleteResolverOpts } from './delete.resolver';
import { BaseResolverOptions } from '../decorators/resolver-method.decorator';
import { mergeBaseResolverOpts } from '../common';
import { RelatableOpts } from './relations/relations.resolver';

export interface CRUDResolverOpts<
  DTO,
  C = DeepPartial<DTO>,
  U = DeepPartial<DTO>,
  R extends ReadResolverOpts<DTO> = ReadResolverOpts<DTO>,
  PS extends PagingStrategies = PagingStrategies.CURSOR,
> extends BaseResolverOptions,
    Pick<ConnectionOptions, 'enableTotalCount'> {
  /**
   * The DTO that should be used as input for create endpoints.
   */
  CreateDTOClass?: Class<C>;
  /**
   * The DTO that should be used as input for update endpoints.
   */
  UpdateDTOClass?: Class<U>;
  enableSubscriptions?: boolean;
  pagingStrategy?: PS;
  enableAggregate?: boolean;
  create?: CreateResolverOpts<DTO, C>;
  read?: R;
  update?: UpdateResolverOpts<DTO, U>;
  delete?: DeleteResolverOpts<DTO>;
  referenceBy?: ReferenceResolverOpts;
  aggregate?: AggregateResolverOpts;
}

export interface CRUDResolver<
  DTO,
  C,
  U,
  R extends ReadResolverOpts<DTO>,
  QS extends QueryService<DTO, C, U> = QueryService<DTO, C, U>,
> extends CreateResolver<DTO, C, QS>,
    ReadResolverFromOpts<DTO, R, QS>,
    UpdateResolver<DTO, U, QS>,
    DeleteResolver<DTO, QS>,
    AggregateResolver<DTO, QS> {}

function extractRelatableOpts<DTO>(
  opts: CRUDResolverOpts<DTO, unknown, unknown, ReadResolverOpts<DTO>, PagingStrategies>,
): RelatableOpts {
  const { enableTotalCount, enableAggregate } = opts;
  return mergeBaseResolverOpts<RelatableOpts>({ enableAggregate, enableTotalCount }, opts);
}

function extractAggregateResolverOpts<DTO>(
  opts: CRUDResolverOpts<DTO, unknown, unknown, ReadResolverOpts<DTO>, PagingStrategies>,
): AggregateResolverOpts {
  const { enableAggregate, aggregate } = opts;
  return mergeBaseResolverOpts<AggregateResolverOpts>({ enabled: enableAggregate, ...aggregate }, opts);
}

function extractCreateResolverOpts<DTO, C>(
  opts: CRUDResolverOpts<DTO, C, unknown, ReadResolverOpts<DTO>, PagingStrategies>,
): CreateResolverOpts<DTO, C> {
  const { CreateDTOClass, enableSubscriptions, create } = opts;
  return mergeBaseResolverOpts<CreateResolverOpts<DTO, C>>({ CreateDTOClass, enableSubscriptions, ...create }, opts);
}

function extractReadResolverOpts<DTO, R extends ReadResolverOpts<DTO>, PS extends PagingStrategies>(
  opts: CRUDResolverOpts<DTO, unknown, unknown, R, PagingStrategies>,
): MergePagingStrategyOpts<DTO, R, PS> {
  const { enableTotalCount, pagingStrategy, read } = opts;
  return mergeBaseResolverOpts(
    { enableTotalCount, pagingStrategy, ...read } as MergePagingStrategyOpts<DTO, R, PS>,
    opts,
  );
}

function extractUpdateResolverOpts<DTO, U>(
  opts: CRUDResolverOpts<DTO, unknown, U, ReadResolverOpts<DTO>, PagingStrategies>,
): UpdateResolverOpts<DTO, U> {
  const { UpdateDTOClass, enableSubscriptions, update } = opts;
  return mergeBaseResolverOpts<UpdateResolverOpts<DTO, U>>({ UpdateDTOClass, enableSubscriptions, ...update }, opts);
}

function extractDeleteResolverOpts<DTO>(
  opts: CRUDResolverOpts<DTO, unknown, unknown, ReadResolverOpts<DTO>, PagingStrategies>,
): DeleteResolverOpts<DTO> {
  const { enableSubscriptions, delete: deleteArgs } = opts;
  return mergeBaseResolverOpts<DeleteResolverOpts<DTO>>({ enableSubscriptions, ...deleteArgs }, opts);
}

/**
 * Factory to create a resolver that includes all CRUD methods from [[CreateResolver]], [[ReadResolver]],
 * [[UpdateResolver]], and [[DeleteResolver]].
 *
 * ```ts
 * import { CRUDResolver } from '@nestjs-query/query-graphql';
 * import { Resolver } from '@nestjs/graphql';
 * import { TodoItemDTO } from './dto/todo-item.dto';
 * import { TodoItemService } from './todo-item.service';
 *
 * @Resolver()
 * export class TodoItemResolver extends CRUDResolver(TodoItemDTO) {
 *   constructor(readonly service: TodoItemService) {
 *     super(service);
 *   }
 * }
 * ```
 * @param DTOClass - The DTO Class that the resolver is for. All methods will use types derived from this class.
 * @param opts - Options to customize the resolver.
 */
// eslint-disable-next-line @typescript-eslint/no-redeclare -- intentional
export const CRUDResolver = <
  DTO,
  C = DeepPartial<DTO>,
  U = DeepPartial<DTO>,
  R extends ReadResolverOpts<DTO> = ReadResolverOpts<DTO>,
  PS extends PagingStrategies = PagingStrategies.CURSOR,
>(
  DTOClass: Class<DTO>,
  opts: CRUDResolverOpts<DTO, C, U, R, PS> = {},
): ResolverClass<DTO, QueryService<DTO, C, U>, CRUDResolver<DTO, C, U, MergePagingStrategyOpts<DTO, R, PS>>> => {
  const referencable = Referenceable(DTOClass, opts.referenceBy ?? {});
  const relatable = Relatable(DTOClass, extractRelatableOpts(opts));
  const aggregateable = Aggregateable(DTOClass, extractAggregateResolverOpts(opts));
  const creatable = Creatable(DTOClass, extractCreateResolverOpts(opts));
  const readable = Readable(DTOClass, extractReadResolverOpts(opts));
  const updatable = Updateable(DTOClass, extractUpdateResolverOpts(opts));
  const deleteResolver = DeleteResolver(DTOClass, extractDeleteResolverOpts(opts));

  return referencable(relatable(aggregateable(creatable(readable(updatable(deleteResolver))))));
};
