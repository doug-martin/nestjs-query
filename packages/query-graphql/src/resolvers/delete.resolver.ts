// eslint-disable-next-line max-classes-per-file
import { Class, DeleteManyResponse, Filter, mergeFilter, QueryService } from '@nestjs-query/core';
import omit from 'lodash.omit';
import { Args, ArgsType, InputType, ObjectType, PartialType, Resolver } from '@nestjs/graphql';
import { HookTypes } from '../hooks';
import { DTONames, getDTONames } from '../common';
import { EventType, getDTOEventName } from '../subscription';
import { BaseServiceResolver, ResolverClass, ServiceResolver, SubscriptionResolverOpts } from './resolver.interface';
import {
  DeleteManyInputType,
  DeleteManyResponseType,
  DeleteOneInputType,
  MutationArgsType,
  SubscriptionArgsType,
  SubscriptionFilterInputType,
} from '../types';
import { MutationHookArgs, ResolverMutation, ResolverSubscription, AuthorizerFilter } from '../decorators';
import { createSubscriptionFilter, getSubscriptionEventName } from './helpers';
import { AuthorizerInterceptor, HookInterceptor } from '../interceptors';
import { OperationGroup } from '../auth';

export type DeletedEvent<DTO> = { [eventName: string]: DTO };
export interface DeleteResolverOpts<DTO> extends SubscriptionResolverOpts {
  /**
   * ArgsType for deleteOne mutation.
   */
  DeleteOneInput?: Class<DeleteOneInputType>;
  /**
   * ArgsType for deleteMany mutation.
   */
  DeleteManyInput?: Class<DeleteManyInputType<DTO>>;
}

export interface DeleteResolver<DTO, QS extends QueryService<DTO, unknown, unknown>> extends ServiceResolver<DTO, QS> {
  deleteOne(input: MutationArgsType<DeleteOneInputType>, authorizeFilter?: Filter<DTO>): Promise<Partial<DTO>>;

  deleteMany(
    input: MutationArgsType<DeleteManyInputType<DTO>>,
    authorizeFilter?: Filter<DTO>,
  ): Promise<DeleteManyResponse>;

  deletedOneSubscription(
    input?: SubscriptionArgsType<DTO>,
    authorizeFilter?: Filter<DTO>,
  ): AsyncIterator<DeletedEvent<Partial<DTO>>>;

  deletedManySubscription(authorizeFilter?: Filter<DTO>): AsyncIterator<DeletedEvent<DeleteManyResponse>>;
}

/** @internal */
const defaultDeleteManyInput = <DTO>(dtoNames: DTONames, DTOClass: Class<DTO>): Class<DeleteManyInputType<DTO>> => {
  const { pluralBaseName } = dtoNames;
  @InputType(`DeleteMany${pluralBaseName}Input`)
  class DM extends DeleteManyInputType(DTOClass) {}
  return DM;
};

/** @internal */
const defaultDeleteOneInput = <DTO>(dtoNames: DTONames, DTOClass: Class<DTO>): Class<DeleteOneInputType> => {
  const { baseName } = dtoNames;
  @InputType(`DeleteOne${baseName}Input`)
  class DM extends DeleteOneInputType(DTOClass) {}
  return DM;
};

/**
 * @internal
 * Mixin to add `delete` graphql endpoints.
 */
export const Deletable =
  <DTO, QS extends QueryService<DTO, unknown, unknown>>(DTOClass: Class<DTO>, opts: DeleteResolverOpts<DTO>) =>
  <B extends Class<ServiceResolver<DTO, QS>>>(BaseClass: B): Class<DeleteResolver<DTO, QS>> & B => {
    const dtoNames = getDTONames(DTOClass, opts);
    const { baseName, pluralBaseName } = dtoNames;
    const enableSubscriptions = opts.enableSubscriptions === true;
    const enableOneSubscriptions = opts.one?.enableSubscriptions ?? enableSubscriptions;
    const enableManySubscriptions = opts.many?.enableSubscriptions ?? enableSubscriptions;
    const deletedOneEvent = getDTOEventName(EventType.DELETED_ONE, DTOClass);
    const deletedManyEvent = getDTOEventName(EventType.DELETED_MANY, DTOClass);
    const {
      DeleteOneInput = defaultDeleteOneInput(dtoNames, DTOClass),
      DeleteManyInput = defaultDeleteManyInput(dtoNames, DTOClass),
    } = opts;
    const deleteOneMutationName = opts.one?.name ?? `deleteOne${baseName}`;
    const deleteManyMutationName = opts.many?.name ?? `deleteMany${pluralBaseName}`;
    const DMR = DeleteManyResponseType();

    const commonResolverOpts = omit(opts, 'dtoName', 'one', 'many', 'DeleteOneInput', 'DeleteManyInput');

    @ObjectType(`${baseName}DeleteResponse`)
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    class DeleteOneResponse extends PartialType(DTOClass, ObjectType) {}

    @ArgsType()
    class DO extends MutationArgsType(DeleteOneInput) {}

    @ArgsType()
    class DM extends MutationArgsType(DeleteManyInput) {}

    @InputType(`DeleteOne${baseName}SubscriptionFilterInput`)
    class SI extends SubscriptionFilterInputType(DTOClass) {}

    @ArgsType()
    class DOSA extends SubscriptionArgsType(SI) {}

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const deleteOneSubscriptionFilter = createSubscriptionFilter(SI, deletedOneEvent);

    @Resolver(() => DTOClass, { isAbstract: true })
    class DeleteResolverBase extends BaseClass {
      @ResolverMutation(
        () => DeleteOneResponse,
        { name: deleteOneMutationName },
        commonResolverOpts,
        { interceptors: [HookInterceptor(HookTypes.BEFORE_DELETE_ONE, DTOClass), AuthorizerInterceptor(DTOClass)] },
        opts.one ?? {},
      )
      async deleteOne(
        @MutationHookArgs() input: DO,
        @AuthorizerFilter({
          operationGroup: OperationGroup.DELETE,
          many: false,
        })
        authorizeFilter?: Filter<DTO>,
      ): Promise<Partial<DTO>> {
        const deletedResponse = await this.service.deleteOne(input.input.id, { filter: authorizeFilter ?? {} });
        if (enableOneSubscriptions) {
          await this.publishDeletedOneEvent(deletedResponse, authorizeFilter);
        }
        return deletedResponse;
      }

      @ResolverMutation(
        () => DMR,
        { name: deleteManyMutationName },
        commonResolverOpts,
        { interceptors: [HookInterceptor(HookTypes.BEFORE_DELETE_MANY, DTOClass), AuthorizerInterceptor(DTOClass)] },
        opts.many ?? {},
      )
      async deleteMany(
        @MutationHookArgs() input: DM,
        @AuthorizerFilter({
          operationGroup: OperationGroup.DELETE,
          many: true,
        })
        authorizeFilter?: Filter<DTO>,
      ): Promise<DeleteManyResponse> {
        const deleteManyResponse = await this.service.deleteMany(
          mergeFilter(input.input.filter, authorizeFilter ?? {}),
        );
        if (enableManySubscriptions) {
          await this.publishDeletedManyEvent(deleteManyResponse, authorizeFilter);
        }
        return deleteManyResponse;
      }

      async publishDeletedOneEvent(dto: DeleteOneResponse, authorizeFilter?: Filter<DTO>): Promise<void> {
        if (this.pubSub) {
          const eventName = getSubscriptionEventName(deletedOneEvent, authorizeFilter);
          await this.pubSub.publish(eventName, { [deletedOneEvent]: dto });
        }
      }

      async publishDeletedManyEvent(dmr: DeleteManyResponse, authorizeFilter?: Filter<DTO>): Promise<void> {
        if (this.pubSub) {
          const eventName = getSubscriptionEventName(deletedManyEvent, authorizeFilter);
          await this.pubSub.publish(eventName, { [deletedManyEvent]: dmr });
        }
      }

      @ResolverSubscription(
        () => DeleteOneResponse,
        { name: deletedOneEvent, filter: deleteOneSubscriptionFilter },
        commonResolverOpts,
        {
          enableSubscriptions: enableOneSubscriptions,
        },
      )
      // input required so graphql subscription filtering will work.
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      deletedOneSubscription(
        @Args() input?: DOSA,
        @AuthorizerFilter({ operationGroup: OperationGroup.DELETE, many: false })
        authorizeFilter?: Filter<DTO>,
      ): AsyncIterator<DeletedEvent<DeleteOneResponse>> {
        if (!enableOneSubscriptions || !this.pubSub) {
          throw new Error(`Unable to subscribe to ${deletedOneEvent}`);
        }
        const eventName = getSubscriptionEventName(deletedOneEvent, authorizeFilter);
        return this.pubSub.asyncIterator(eventName);
      }

      @ResolverSubscription(() => DMR, { name: deletedManyEvent }, commonResolverOpts, {
        enableSubscriptions: enableManySubscriptions,
      })
      deletedManySubscription(
        @AuthorizerFilter({ operationGroup: OperationGroup.DELETE, many: true })
        authorizeFilter?: Filter<DTO>,
      ): AsyncIterator<DeletedEvent<DeleteManyResponse>> {
        if (!enableManySubscriptions || !this.pubSub) {
          throw new Error(`Unable to subscribe to ${deletedManyEvent}`);
        }
        const eventName = getSubscriptionEventName(deletedManyEvent, authorizeFilter);
        return this.pubSub.asyncIterator(eventName);
      }
    }
    return DeleteResolverBase;
  };
// eslint-disable-next-line @typescript-eslint/no-redeclare -- intentional
export const DeleteResolver = <
  DTO,
  QS extends QueryService<DTO, unknown, unknown> = QueryService<DTO, unknown, unknown>,
>(
  DTOClass: Class<DTO>,
  opts: DeleteResolverOpts<DTO> = {},
): ResolverClass<DTO, QS, DeleteResolver<DTO, QS>> => Deletable<DTO, QS>(DTOClass, opts)(BaseServiceResolver);
