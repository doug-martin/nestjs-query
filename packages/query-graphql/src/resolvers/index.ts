export { CRUDResolverOpts, CRUDResolver } from './crud.resolver';
export { CreateResolver, CreateResolverOpts } from './create.resolver';
export { ReadResolver, ReadResolverOpts } from './read.resolver';
export { UpdateResolver, UpdateResolverOpts } from './update.resolver';
export { DeleteResolver, DeleteResolverOpts } from './delete.resolver';
export { ReferenceResolver, ReferenceResolverOpts } from './reference.resolver';
export { Relatable, ResolverRelation, ResolverRelationReference } from './relations';
export { ResolverOpts } from './resolver.interface';
export { FederationResolver } from './federation';
