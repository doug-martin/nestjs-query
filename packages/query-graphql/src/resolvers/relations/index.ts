export { Relatable } from './relations.resolver';
export { ReadRelationsResolver } from './read-relations.resolver';
export { UpdateRelationsResolver } from './update-relations.resolver';
export { RemoveRelationsResolver } from './remove-relations.resolver';
export { ReferencesRelationsResolver } from './references-relation.resolver';
export { AggregateRelationsResolver } from './aggregate-relations.resolver';
export {
  RelationTypeMap,
  ReferencesOpts,
  ResolverRelationReference,
  RelationsOpts,
  ResolverRelation,
} from './relations.interface';
