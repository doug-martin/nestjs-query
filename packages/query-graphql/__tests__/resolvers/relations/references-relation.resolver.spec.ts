import { Resolver, Query } from '@nestjs/graphql';
import { ReferencesOpts, ReferencesRelationsResolver } from '../../../src/resolvers/relations';
import {
  generateSchema,
  createResolverFromNest,
  TestResolverDTO,
  TestService,
  TestRelationDTO,
} from '../../__fixtures__';

@Resolver(() => TestResolverDTO)
class TestResolver extends ReferencesRelationsResolver(TestResolverDTO, {
  reference: { DTO: TestRelationDTO, keys: { id: 'stringField' } },
}) {
  constructor(service: TestService) {
    super(service);
  }
}

describe('ReferencesRelationMixin', () => {
  const expectResolverSDL = async (opts?: ReferencesOpts<TestResolverDTO>) => {
    @Resolver(() => TestResolverDTO)
    class TestSDLResolver extends ReferencesRelationsResolver(TestResolverDTO, opts ?? {}) {
      @Query(() => TestResolverDTO)
      test(): TestResolverDTO {
        return { id: '1', stringField: 'foo' };
      }
    }

    const schema = await generateSchema([TestSDLResolver]);
    expect(schema).toMatchSnapshot();
  };
  it('should not add reference methods if references empty', () => expectResolverSDL());

  it('should use the add the reference if provided', () =>
    expectResolverSDL({ reference: { DTO: TestRelationDTO, keys: { id: 'stringField' }, dtoName: 'Test' } }));

  it('should set the field to nullable if set to true', () =>
    expectResolverSDL({ reference: { DTO: TestRelationDTO, keys: { id: 'stringField' }, nullable: true } }));

  it('should return a references type from the passed in dto', async () => {
    const { resolver } = await createResolverFromNest(TestResolver);
    const dto: TestResolverDTO = {
      id: 'id-1',
      stringField: 'reference-id-1',
    };
    // @ts-ignore
    // eslint-disable-next-line @typescript-eslint/no-unsafe-call
    const result = await resolver.referenceReference(dto);
    // eslint-disable-next-line @typescript-eslint/naming-convention
    return expect(result).toEqual({ __typename: 'Reference', id: dto.stringField });
  });
});
