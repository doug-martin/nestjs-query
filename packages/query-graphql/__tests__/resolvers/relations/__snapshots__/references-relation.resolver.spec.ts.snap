// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReferencesRelationMixin should not add reference methods if references empty 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type Query {
  test: TestResolverDTO!
}

`;

exports[`ReferencesRelationMixin should set the field to nullable if set to true 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
  reference: TestRelationDTO
}

type TestRelationDTO {
  id: ID!
  testResolverId: String!
}

type Query {
  test: TestResolverDTO!
}

`;

exports[`ReferencesRelationMixin should use the add the reference if provided 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
  test: TestRelationDTO!
}

type TestRelationDTO {
  id: ID!
  testResolverId: String!
}

type Query {
  test: TestResolverDTO!
}

`;
